File: doc-examples/hibernate-reactive-example-java/src/main/java/example/ProductManager.java

package example;

import io.micronaut.data.connection.reactive.ReactorConnectionOperations;
import io.micronaut.transaction.reactive.ReactorReactiveTransactionOperations;
import jakarta.inject.Singleton;
import org.hibernate.reactive.stage.Stage;
import reactor.core.publisher.Mono;

@Singleton
public class ProductManager {

    private final ReactorConnectionOperations<Stage.Session> connectionOperations;
    private final ReactorReactiveTransactionOperations<Stage.Session> transactionOperations;

    public ProductManager(ReactorConnectionOperations<Stage.Session> connectionOperations,
                          ReactorReactiveTransactionOperations<Stage.Session> transactionOperations) {
        this.connectionOperations = connectionOperations;
        this.transactionOperations = transactionOperations;
    }

    Mono<Product> save(String name, Manufacturer manufacturer) {
        return transactionOperations.withTransactionMono(status -> {
            final Product product = new Product(name, manufacturer);
            return Mono.fromCompletionStage(() -> status.getConnection().persist(product)).thenReturn(product);
        });
    }

    Mono<Product> find(String name) {
        return transactionOperations.withTransactionMono(status -> Mono.fromCompletionStage(status.getConnection().createQuery("from Product p where p.name = :name", Product.class)
                .setParameter("name", name)
                .getSingleResult())
        );
    }
}

File: doc-examples/hibernate-reactive-example-kotlin/src/main/kotlin/example/ProductManager.kt

package example

import io.micronaut.transaction.reactive.ReactiveTransactionStatus
import io.micronaut.transaction.reactive.ReactorReactiveTransactionOperations
import jakarta.inject.Singleton
import org.hibernate.reactive.stage.Stage
import reactor.core.publisher.Mono

@Singleton
class ProductManager(
        private val transactionManager: ReactorReactiveTransactionOperations<Stage.Session>
) // <1>
{

    fun save(name: String, manufacturer: Manufacturer): Mono<Product> {
        return transactionManager.withTransactionMono { status: ReactiveTransactionStatus<Stage.Session> -> // <2>
            val product = Product(null, name, manufacturer)
            Mono.fromCompletionStage {
                status.connection.persist(product)
            }.thenReturn(product)
        }
    }

    fun find(name: String): Mono<Product> {
        return transactionManager.withTransactionMono { status: ReactiveTransactionStatus<Stage.Session> -> // <3>
            Mono.fromCompletionStage(
                status.connection.createQuery(
                    "from Product p where p.name = :name",
                    Product::class.java
                )
                    .setParameter("name", name)
                    .singleResult
            )
        }
    }
}
